test/libraries/encoders/SwapEncoder.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.8.26;
+
+import {Vm} from "forge-std/Test.sol";
+
+import {IERC20, IVault, GPv2Order, GPv2Trade, GPv2Signing} from "src/contracts/GPv2Settlement.sol";
+
+import {Sign} from "../Sign.sol";
+import {Trade} from "../Trade.sol";
+import {Registry, TokenRegistry} from "./TokenRegistry.sol";
+
+library SwapEncoder {
+    using Trade for GPv2Order.Data;
+    using Sign for Vm;
+    using TokenRegistry for TokenRegistry.State;
+    using TokenRegistry for Registry;
+
+    struct Swap {
+        bytes32 poolId;
+        IERC20 assetIn;
+        IERC20 assetOut;
+        uint256 amount;
+        bytes userData;
+    }
+
+    struct EncodedSwap {
+        IVault.BatchSwapStep[] swaps;
+        IERC20[] tokens;
+        GPv2Trade.Data trade;
+    }
+
+    struct State {
+        Registry tokenRegistry;
+        IVault.BatchSwapStep[] steps;
+        GPv2Trade.Data trade;
+    }
+
+    bytes32 internal constant STATE_STORAGE_SLOT = keccak256("SwapEncoder.storage");
+
+    /// @dev Make a new swap encoder derived from the specified encoder ID
+    function makeSwapEncoder(address encoder) internal returns (State storage state) {
+        bytes32 slot = keccak256(abi.encodePacked(STATE_STORAGE_SLOT, encoder));
+        assembly {
+            state.slot := slot
+        }
+
+        setTokenRegistry(state, Registry.wrap(keccak256(abi.encode(slot))));
+    }
```

Comment:
Same discussion as above. What is the `encoder` address? Can it be replaced with a nonce? The only reason I can think of why we shouldn't do it is that we want to deterministically generate an encoder starting from an `address` rather than creating the object and moving it around. The latter use case would be very complex and I'd expect all cases where we use an encoder to be self-contained.

Reply:
Done.